### PR TITLE
Making IgnoredInvites mostly sync.

### PR DESCRIPTION
**Pull request by Yoric**
Experience and feedback from https://github.com/matrix-org/matrix-react-sdk/pull/9255 indicates that we cannot integrate an async `IgnoredInvites.getRuleForInvite`. This PR:

- rewrites `getRuleForInvite` to be sync;
- adds some locking within `IgnoredInvites` to avoid race conditions noticed by @t3chguy.

## Checklist

* [X] Tests written for new code (and old code if feasible)
* [X] Linter and other CI checks pass
* [X] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Type: task

### CODE REVIEW

File: src/models/invites-ignorer.ts
Comment (t3chguy) [inline src/models/invites-ignorer.ts:253]: Does this need to lock? We could instead return the same promise, no?

Comment (Yoric) [reply src/models/invites-ignorer.ts:253]: Which promise? That one?

```ts
this.client.createRoom({
                name: "Individual Policy Room",
                preset: Preset.PrivateChat,
            })
```
We could probably store it and use it as a makeshift lock, but it feels more fragile/harder to read. Since we don't expect that this block will be performance-critical (in most runs, it should be executed either never or once), I'd favour the most readable solution, whichever it is.

Comment (t3chguy) [reply src/models/invites-ignorer.ts:253]: We have that pattern in other places - https://github.com/matrix-org/matrix-js-sdk/blob/8aee884d03d79934e8944b05c22643c1bb480a64/src/models/room.ts#L409-L427 

I wouldn't suggest just storing the promise of the createRoom call, but rather this whole method by wrapping it.

Comment (Yoric) [reply src/models/invites-ignorer.ts:253]: Yeah, that would work but it feels a bit more brittle in presence of exception.

Your call.

Comment (Yoric) [reply src/models/invites-ignorer.ts]: Done.
As always, documenting concurrency is an interesting pastime :)

File: src/models/invites-ignorer.ts
Comment (t3chguy) [inline src/models/invites-ignorer.ts:360]: Not sure a lock is better here, as it yield N `setAccountData` calls which don't de-dup lower down so the promise-d/deferred approach probably works better here also

Comment (Yoric) [reply src/models/invites-ignorer.ts:360]: I'm not sure what you have in mind.

Here, if the code is called N times, it's because we have N different things to write, so writing N times kinda makes sense. I suppose that I could accumulate them into a single call to `setAccountData` by storing the changes somewhere, to decrease network I/O in the unlikely case `withIgnoreInvitesPolicies` is called concurrently?

This would definitely complicate the code, though, and I don't really see what it optimizes for.

Or am I misunderstanding what you have in mind?

Comment (t3chguy) [reply src/models/invites-ignorer.ts:360]: This lock is just providing linearisation, which can be accomplished with a Promise. 

```
withIgnoreInvitesPoliciesPromise = Promise.resolve<void>();
...
await this.withIgnoreInvitesPoliciesPromise;
...
this.withIgnoreInvitesPoliciesPromise = this.client.setAccountData(...);
```

Comment (t3chguy) [reply src/models/invites-ignorer.ts:360]: Also there are missing types in `getPoliciesAndIgnoreInvitesPolicies`

File: src/models/invites-ignorer.ts
Comment (t3chguy) [inline src/models/invites-ignorer.ts:79]: This shouldn't be underscored as per the code style

File: src/models/invites-ignorer.ts
Comment (t3chguy) [inline src/models/invites-ignorer.ts:84]: This shouldn't be underscored as per the code style

Comment (t3chguy) [reply src/models/invites-ignorer.ts:84]: There should also be a blank line between this and the constructor

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:360]: > Also there are missing types in `getPoliciesAndIgnoreInvitesPolicies`

Not following. Do you wish me to give names to `{[key: string]: any}` types?

**Comment by Yoric** [comment]
@t3chguy Is this any better? The `any` code makes it clearer what data has been validated and what data hasn't.

**Comment by t3chguy** [comment]
@Yoric sorry going to have to bump this back to the queue due to other obligations

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:102]: shouldn't this be part of the description? At the moment it's part of the `@return`

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:275]: Should this be translated?

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:272]: typically our pattern is `= new Promise((resolve, reject) => { ... });` instead

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:344]: this needs a type - what is it?

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:365]: these also need types

File: yarn.lock
Comment (turt2live) [inline yarn.lock:1413]: this seems unrelated?

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:162]: this is a breaking change because the class is not marked as unstable

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:83]: these look like functions if the reader isn't extremely careful: can we call them something like `targetRoomLock` instead?

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:143]: why is this on two lines? It's difficult to read this way :(

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:139]: we don't await it though?

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:240]: this is the wrong field name, likewise throughout the rest of the comment block

File: src/models/invites-ignorer.ts
Comment (turt2live) [inline src/models/invites-ignorer.ts:238]: This comment (and the lines which follow) aren't really describing anything: it's just saying what we're about to do, not why it's important to have a lock. We know what it's about to do: there's code to tell us :)

Comment (turt2live) [reply src/models/invites-ignorer.ts:238]: We should also describe why we don't need cross-client locking, considering this code is seemingly important.

It's further a bad smell that we need a lock at all - why would this code be called multiple times?

**Review by turt2live — CHANGES_REQUESTED** [review]
sorry, I'm coming at this blind with no real idea what troubles you're facing on the react-sdk side: it may be worth a visit to Element Web Internal to talk about those challenges instead, as making things synchronous can be annoyingly more difficult than understanding the async code path.

This is also much harder to review because there's no MSC reference within the area of the diff - it's hard to see what the underlying data structures are without such a reference.

**Comment by Yoric** [comment]
> sorry, I'm coming at this blind with no real idea what troubles you're facing on the react-sdk side: it may be worth a visit to Element Web Internal to talk about those challenges instead, as making things synchronous can be annoyingly more difficult than understanding the async code path.

Well, I've been asked by my react-sdk reviewer to make things sync :)

> This is also much harder to review because there's no MSC reference within the area of the diff - it's hard to see what the underlying data structures are without such a reference.

Should be https://github.com/matrix-org/matrix-spec-proposals/pull/3847, iirc.

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:275]: Oh, good point. Is there any way to do that in matrix-js-sdk? Or should I somehow pass that string from react-sdk?

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:344]: > this needs a type - what is it?

It's data coming from Matrix, therefore untrusted. Until I have checked it (which is what the caller functions do), I have no clue. The `any` is there specifically to mark that it's untrusted. Is there a better way to write it down?

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:162]: Well, this is a breaking change on my code that afaict isn't used anywhere yet, because the react-sdk patch was rejected because of `async` :)

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:143]: :shrug: Different codebases, different conventions. I'll adapt :)

File: src/models/invites-ignorer.ts
Comment (t3chguy) [reply src/models/invites-ignorer.ts:162]: The issue is this wasn't marked experimental and there's 70+ dependants on matrix-js-sdk, any of which could use a release where the API was synchronous and then have major issues when updating within the same semver range, so it is a breaking change.

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:238]: > This comment (and the lines which follow) aren't really describing anything: it's just saying what we're about to do, not why it's important to have a lock. We know what it's about to do: there's code to tell us :)

...because of race conditions? I'm reading, modifying, writing some user account data, that's inherently racy.

> We should also describe why we don't need cross-client locking, considering this code is seemingly important.

Wait, we have cross-client locking? Where do I find the documentation on these?

> It's further a bad smell that we need a lock at all - why would this code be called multiple times?

I come from a safety background. Any code is considered to be called multiple times unless proven otherwise :)

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:272]: With two calls to `then` after the `Promise`? Or is it an `async` `(resolve, reject) => { ... }`?

I've tried rewriting this call to use the `Promise` constructor but it didn't really feel readable.

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:365]: Same response, sadly. If there is a better way to mark untrusted-deserialized-json than `any`, I'll be happy to use it!

File: src/models/invites-ignorer.ts
Comment (Yoric) [reply src/models/invites-ignorer.ts:162]: @t3chguy So how do I handle this change? Should I add a new method instead of replacing the old one?

Comment (t3chguy) [reply src/models/invites-ignorer.ts:158]: Ideally yes, or mark it as X-Breaking-Change but then it might get held until we have a few to go together for a major version bump

**Review by turt2live — CHANGES_REQUESTED** [review]
this appears to have failing checks, and it's still not really clear why this PR needs to exist. Usually we're heavily in support of async code, so I'm surprised we want to make it sync.

It's also hard to re-review this at the moment: all the conversations have been left unresolved, making it unclear which need further comment.

**Comment by Yoric** [comment]
> this appears to have failing checks, and it's still not really clear why this PR needs to exist. Usually we're heavily in support of async code, so I'm surprised we want to make it sync.
> 
> It's also hard to re-review this at the moment: all the conversations have been left unresolved, making it unclear which need further comment.

To clarify: I re-requested review because at least 3 conversations are waiting for a response on your part. I realize that I should have cleared up the resolved conversations, my bad.

Also, I'm making the code sync because @t3chguy requested that I do so [here](https://github.com/matrix-org/matrix-react-sdk/pull/9255#discussion_r964931213).

**Comment by florianduros** [comment]
Hi!

The PR is 2 years old and stuck. 
I'm closing it. Feel free to reopen it if you want to continue to work on it.

**Pull request closed without merging**
